File: cpp/1137_N-th_Tribonacci_Number.cpp
class Solution {
    map<int, int> memo = {{0, 0}, {1, 1}, {2, 1}};
public:
    int tribonacci(int n) {
        if (memo.find(n) == memo.end()) {
            memo[n] = tribonacci(n - 1)
                + tribonacci(n - 2) + tribonacci(n - 3);
        }
        return memo[n];
    }
};

File: java/1137_N-th_Tribonacci_Number.java
class Solution {
    HashMap<Integer, Integer> memo = new HashMap<Integer, Integer>();

    public int tribonacci(int n) {

        if (n == 0 || n == 1) {
            return n;
        }
        if (n == 2) {
            return 1;
        }
        if (!memo.containsKey(n)) {
            memo.put(n, tribonacci(n - 1) + tribonacci(n - 2) + tribonacci(n - 3));
        }

        return memo.get(n);
    }
}
